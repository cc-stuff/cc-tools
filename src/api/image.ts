import * as path from 'path';
import * as fs from 'fs-extra';
import {formatFileSize} from "../common/formatFileSize";
import {assertIsString} from "../common/assertIsString";
import {resolvePath} from "../common/resolvePath";

interface FSNodeAttributes {
	creation: number;
	modification: number;
}

interface FSNode {
	attributes: FSNodeAttributes;
}

interface FSFile extends FSNode {
	b64: string;
}

interface FSDirectory extends FSNode {
	children: string[];
}

interface Image {
	[key: string]: FSFile | FSDirectory;
}

function readDir(rootDir: string, dir: string, outImage: Image): string[] {
	const files = fs.readdirSync(path.resolve(rootDir, dir));

	for (const fileName of files) {
		const relFileName = `${dir}/${fileName}`;
		const absFileName = path.resolve(rootDir, relFileName);
		const fileKey = `computer[0].files[${relFileName.slice(2)}]`;

		const fileStats = fs.statSync(absFileName);

		const attributes = {
			creation: Date.now(),
			modification: Date.now(),
		}

		if (fileStats.isDirectory()) {
			const childFiles = readDir(rootDir, relFileName, outImage);

			outImage[fileKey] = {
				attributes,
				children: childFiles,
			}
		} else {
			outImage[fileKey] = {
				attributes,
				b64: fs.readFileSync(absFileName).toString("base64"),
			}
		}
	}

	return files;
}

function flattenImage(image: Image): [string, string][] {
	const result = [];

	for (const key of Object.keys(image)) {
		const entry = image[key];

		result.push([`${key}.attributes`, JSON.stringify(entry.attributes)]);

		if ((entry as FSDirectory).children) {
			result.push([`${key}.children`, JSON.stringify((entry as FSDirectory).children)]);
		} else {
			result.push([`${key}.b64`, JSON.stringify((entry as FSFile).b64)]);
		}
	}

	return result;
}

function writeImage(flat: [string, string][], outDir: string): number {
	let extractedCount = 0;

	for (const entry of flat) {
		const [key, value] = entry;

		const fileName = /^computer\[0]\.files\[(.*)]/.exec(key)[1];
		const fileNameAbs = resolvePath(fileName, outDir);

		// Forcing the path
		fs.mkdirpSync(path.dirname(fileNameAbs));

		// Creating a file
		if (key.endsWith(".b64")) {
			console.log(`Extracting ${fileNameAbs}`);

			fs.writeFileSync(fileNameAbs, Buffer.from(value, "base64").toString("utf-8"));
			extractedCount++;
		}
	}

	return extractedCount;
}

export interface PackImageOptions {
	args?: {
		folder: string;
		output: string;
		noEmit?: boolean;
	}
}

export function packImage(options: PackImageOptions) {
	const {folder, output} = options.args;

	assertIsString(folder);

	const rootDir = resolvePath(folder);

	if (typeof output === "string" && fs.existsSync(output)) {
		fs.unlinkSync(output);
	}

	const image: Image = {};

	image["computer[0].files[]"] = {
		attributes: {
			creation: 0,
			modification: 0,
		},
		children: readDir(rootDir, ".", image),
	}

	const imageOut = JSON.stringify(flattenImage(image));

	if (!options.args.noEmit) {
		assertIsString(output);

		fs.writeFileSync(output, imageOut, "utf-8");

		console.log('Out image: ', formatFileSize(fs.statSync(output).size));
	}

	return imageOut
}

export interface UnpackImageOptions {
	args?: {
		fs?: string;
		image?: string;
		output: string;
	}
}

export function unpackImage(options: UnpackImageOptions) {
	const {fs: imageFile, image, output} = options.args;

	assertIsString(output);

	const outputAbs = resolvePath(output);

	const rawImage = typeof imageFile === "string"
		? fs.readFileSync(resolvePath(imageFile), "utf-8")
		: image;

	const flatImage: [string, string][] = JSON.parse(rawImage);

	const extractedCount = writeImage(flatImage, outputAbs);

	console.log('Extracted files: ', extractedCount);
}
