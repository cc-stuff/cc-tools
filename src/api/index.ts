export {BundleProjectOptions, bundleProject} from "./project";
export {PackImageOptions, packImage} from "./image";
export {RunOptions, ComputerCraftInstance, run} from "./runner";

export interface UnpackImageOptions {
	args?: {
		image: string;
		output: string;
	}
}

export function unpackImage(options: UnpackImageOptions) {
}

export interface TestOptions {
	args?: {
		project: string;
		testsGrep?: string;
		suitsGrep?: string;
	}
}

export function test(options: TestOptions) {
}
