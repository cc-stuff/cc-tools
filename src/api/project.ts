import * as path from 'path';
import * as fs from 'fs';
import {formatFileSize} from "../common/formatFileSize";
import {assertIsString} from "../common/assertIsString";
import {resolvePath} from "../common/resolvePath";

interface CCBundleConfig {
	entry: string;
	output: string;
	references?: Record<string, string>;
}

interface Project {
	config: CCBundleConfig;
	rootDir: string;
	files: Record<string, File>;
}

interface File {
	source: string;
	path: string;
	moduleName: string;
}

const bundleTemplate = (body, entryModuleName) => `
local __files__ = {}

local function __require__(name)
	return __files__[name]
end

local function __define__(name, impl)
	__files__[name] = impl()
end

${body}

__require__('${entryModuleName}')
`;

const defineTemplate = (name, content) => `
__define__('${name}', function ()
${content}
end)
`;

const requireTemplate = name => `__require__('${name}')`;

function requireFile(project: Project, parent: File, relativeFileName: string): File {
	const absPath = path.resolve(path.dirname(parent.path), relativeFileName) +
		(relativeFileName.endsWith('.lua') ? '' : '.lua');

	if (project.files[absPath]) {
		return project.files[absPath];
	}

	const file: File = {
		source: fs.readFileSync(absPath, 'utf-8'),
		path: absPath,
		moduleName: path
			.relative(project.rootDir, absPath)
			.replace(/\\/g, "/"),
	}

	expandRequires(file, project);

	file.source = defineTemplate(file.moduleName, file.source);
	project.files[absPath] = file;

	return file;
}

function expandRequires(file: File, project: Project) {
	const regexp = /require\s+['"](.*)['"]/gmi;
	const result = [];
	let matches;
	let lastIndex = 0;

	while ((matches = regexp.exec(file.source)) !== null) {
		const requiredFile = requireFile(project, file, matches[1]);

		result.push(file.source.substring(lastIndex, matches.index));
		result.push(requireTemplate(requiredFile.moduleName));

		lastIndex = matches.index + matches[0].length;
	}

	result.push(file.source.substring(lastIndex));

	file.source = result.join('');
}

export interface BundleProjectOptions {
	projectFile?: string;
	args?: {
		entry: string;
		output?: string;
		noEmit?: boolean;
	}
}

export function bundleProject(options: BundleProjectOptions) {
	const {projectFile} = options;
	const {entry: entryArg, output: outputArg} = options.args ?? {};

	// Creating project
	const project: Project = {
		config: {
			entry: "",
			output: "",
		},
		rootDir: process.cwd(),
		files: {},
	}

	// Loading the config from the file
	if (typeof projectFile === "string") {
		const projectFileAbs = resolvePath(projectFile);
		const projectRoot = path.dirname(projectFileAbs);

		project.config = JSON.parse(fs.readFileSync(projectFileAbs, "utf-8"));
		project.config.entry = resolvePath(project.config.entry, projectRoot);
		project.config.output = resolvePath(project.config.output, projectRoot);
	} else {
		assertIsString(entryArg);
		project.config.entry = resolvePath(entryArg);

		if (!options.args.noEmit) {
			assertIsString(outputArg);

			project.config.output = resolvePath(outputArg);
		}
	}

	project.rootDir = path.dirname(project.config.entry);

	// Header stub file
	const headerFile: File = {
		source: '',
		path: path.resolve(project.rootDir, './__header__'),
		moduleName: '',
	}

	// Requiring entry file
	const entryFile = requireFile(project, headerFile, path.basename(project.config.entry));

	// Creating bundle source
	const bundleSource = bundleTemplate(
		Object
			.values(project.files)
			.map(file => file.source)
			.join(''),
		entryFile.moduleName,
	);

	// Validate that it's eiter noEmit or references but never both
	if (options.args.noEmit && project.config.references) {
		throw new Error("noEmit can apply only to projects without references")
	}

	// Writing bundle to file
	if (!options.args.noEmit) {
		fs.writeFileSync(project.config.output, bundleSource, 'utf-8');

		// Bundle size info
		console.log(
			`[${path.basename(project.config.output)}] Out bundle: `,
			formatFileSize(fs.statSync(project.config.output).size),
		);
	}

	// Bundling child projects
	if (project.config.references) {
		for (let reference of Object.values(project.config.references)) {
			reference = resolvePath(reference, project.rootDir);

			bundleProject({projectFile: reference});
		}
	}

	return bundleSource;
}
