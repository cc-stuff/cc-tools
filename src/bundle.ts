import * as path from 'path';
import * as fs from 'fs';
import {argv} from 'yargs';
import * as assert from "assert";
import {formatFileSize} from "./common/formatFileSize";
import {assertIsString} from "./common/assertIsString";

interface Project {
	rootDir: string;
	files: Record<string, File>;
}

interface File {
	source: string;
	path: string;
	moduleName: string;
}

const bundleTemplate = (body, entryModuleName) => `
local __files__ = {}

local function __require__(name)
	return __files__[name]
end

local function __define__(name, impl)
	__files__[name] = impl()
end

${body}

__require__('${entryModuleName}')
`;

const defineTemplate = (name, content) => `
__define__('${name}', function ()
${content}
end)
`;

const requireTemplate = name => `__require__('${name}')`;

function requireFile(project: Project, parent: File, relativeFileName: string): File {
	const absPath = path.resolve(path.dirname(parent.path), relativeFileName) +
		(relativeFileName.endsWith('.lua') ? '' : '.lua');

	if (project.files[absPath]) {
		return project.files[absPath];
	}

	const file: File = {
		source: fs.readFileSync(absPath, 'utf-8'),
		path: absPath,
		moduleName: absPath.substring(project.rootDir.length).replace(/\\/g, '/'),
	}

	expandRequires(file, project);

	file.source = defineTemplate(file.moduleName, file.source);
	project.files[absPath] = file;

	return file;
}

function expandRequires(file: File, project: Project) {
	const regexp = /require\s+['"](.*)['"]/gmi;
	const result = [];
	let matches;
	let lastIndex = 0;

	while ((matches = regexp.exec(file.source)) !== null) {
		const requiredFile = requireFile(project, file, matches[1]);

		result.push(file.source.substring(lastIndex, matches.index));
		result.push(requireTemplate(requiredFile.moduleName));

		lastIndex = matches.index + matches[0].length;
	}

	result.push(file.source.substring(lastIndex));

	file.source = result.join('');
}

const {entry, output} = argv;

assertIsString(entry);
assertIsString(output);

const absEntry = path.isAbsolute(entry)
	? entry
	: path.resolve(process.cwd(), entry);

const project: Project = {
	rootDir: path.dirname(absEntry),
	files: {},
}

const headerFile: File = {
	source: '',
	path: path.resolve(project.rootDir, './__header__'),
	moduleName: '',
}

const entryFile = requireFile(project, headerFile, path.basename(absEntry));

const bundleSource = bundleTemplate(
	Object
		.values(project.files)
		.map(file => file.source)
		.join(''),
	entryFile.moduleName,
);

fs.writeFileSync(output, bundleSource, 'utf-8');

console.log('Out bundle: ', formatFileSize(fs.statSync(output).size));
