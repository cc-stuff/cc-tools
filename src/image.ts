import * as path from 'path';
import * as fs from 'fs';
import {argv} from 'yargs';
import {formatFileSize} from "./common/formatFileSize";
import {assertIsString} from "./common/assertIsString";
import {resolvePath} from "./common/resolvePath";

interface FSNodeAttributes {
	creation: number;
	modification: number;
}

interface FSNode {
	attributes: FSNodeAttributes;
}

interface FSFile extends FSNode {
	b64: string;
}

interface FSDirectory extends FSNode {
	children: string[];
}

interface Image {
	[key: string]: FSFile | FSDirectory;
}

function readDir(rootDir: string, dir: string, outImage: Image): string[] {
	const files = fs.readdirSync(path.resolve(rootDir, dir));

	for (const fileName of files) {
		const relFileName = `${dir}/${fileName}`;
		const absFileName = path.resolve(rootDir, relFileName);
		const fileKey = `computer[0].files[${relFileName.slice(2)}]`;

		const fileStats = fs.statSync(absFileName);

		const attributes = {
			creation: Date.now(),
			modification: Date.now(),
		}

		if (fileStats.isDirectory()) {
			const childFiles = readDir(rootDir, relFileName, outImage);

			outImage[fileKey] = {
				attributes,
				children: childFiles,
			}
		} else {
			outImage[fileKey] = {
				attributes,
				b64: fs.readFileSync(absFileName).toString("base64"),
			}
		}
	}

	return files;
}

function flattenImage(image: Image): [string, string][] {
	const result = [];

	for (const key of Object.keys(image)) {
		const entry = image[key];

		result.push([`${key}.attributes`, JSON.stringify(entry.attributes)]);

		if ((entry as FSDirectory).children) {
			result.push([`${key}.children`, JSON.stringify((entry as FSDirectory).children)]);
		} else {
			result.push([`${key}.b64`, JSON.stringify((entry as FSFile).b64)]);
		}
	}

	return result;
}

const {folder, output} = argv;

assertIsString(folder);
assertIsString(output);

const rootDir = resolvePath(folder);

if (fs.existsSync(output)) {
	fs.unlinkSync(output);
}

const image: Image = {};

image["computer[0].files[]"] = {
	attributes: {
		creation: 0,
			modification: 0,
	},
	children: readDir(rootDir, ".", image),
}

fs.writeFileSync(output, JSON.stringify(flattenImage(image)), "utf-8");

console.log('Out image: ', formatFileSize(fs.statSync(output).size));
