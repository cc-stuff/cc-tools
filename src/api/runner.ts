import {Browser, chromium, Page} from "playwright";
import * as path from 'path';
import * as fs from 'fs';
import {createServer} from "http-server";
import * as http from "http";
import {resolvePath} from "../common/resolvePath";
import {EventEmitter} from "events";
import {packImage, unpackImage} from "./image";

declare const __LS_ITEMS__: [string, string][];

interface Context {
	options: RunOptions;
	browserPromise?: Promise<Browser>;
	imageFileAbs?: string;
	image?: string;
	server?: http.Server;
}

function runChromium(context: Context): Promise<Browser> {
	if (!context.browserPromise) {
		context.browserPromise = chromium.launch({
			headless: false,
		});
	}

	return context.browserPromise;
}

export interface ComputerCraftInstance extends EventEmitter {
	readonly page: Page;

	stop(): Promise<void>;
	waitForStatus(status: "started" | "ended"): Promise<void>;
	getRawImage(): string;
}

async function initializePage(context: Context): Promise<ComputerCraftInstance> {
	const imageSrc = typeof context.imageFileAbs === "string"
		? fs.readFileSync(context.imageFileAbs, "utf-8")
		: context.image;

	const image = JSON.parse(imageSrc);

	const browser = await runChromium(context);

	const page = await browser.newPage();

	const initScript = "(" + function () {
		function patchStartupFile(originalFile?: string): string {
			return btoa(`
cctools = {}

cctools.start = function()
	local writeCommand = function(sCommand)
		local file = fs.open(".ccstatus", "w")
		file.write(sCommand)
		file.close()
	end

	cctools.finish = function()
		os.startTimer(1)
		os.pullEvent()
		writeCommand("ended")
	end

	writeCommand("started")
end

cctools.start()

${originalFile === undefined ? "" : atob(originalFile)}
`);
		}

		const items = __LS_ITEMS__;
		const startupKey = "computer[0].files[startup.lua].b64";
		const statusKey = "computer[0].files[.ccstatus].b64";

		localStorage.clear();

		for (const [key, value] of items) {
			let parsedValue = key.endsWith(".b64")
				? JSON.parse(value)
				: value;

			// Patching startup file
			if (key === startupKey) {
				parsedValue = patchStartupFile(parsedValue);
			}

			localStorage.setItem(key, parsedValue);
		}

		// Hook to clear status
		(window as any).__clearStatus__ = () => {
			const statusElement = document.getElementById("ccstatus");

			if (statusElement) {
				statusElement.parentNode.removeChild(statusElement);
			}
		}

		const originalSetItem = localStorage.setItem;

		// Patching localStorage.set to monitor ComputerCraft FS changes
		localStorage.setItem = function (key: string, value: any) {
			// Displaying status in DOM so that Node process can pick it up
			if (key === statusKey) {
				(window as any).__clearStatus__();

				const statusElement = document.createElement("DIV");
				statusElement.id = "ccstatus";
				statusElement.style.display = "none";
				statusElement.innerText = atob(String(value));

				document.body.appendChild(statusElement);

				console.log("status changed", statusElement.innerText);
			}

			originalSetItem.apply(localStorage, arguments);
		}
	}.toString().replace("__LS_ITEMS__", JSON.stringify(image)) + ")()";

	await page.addInitScript(initScript);

	const instance: ComputerCraftInstance = new EventEmitter() as ComputerCraftInstance;
	const server = await startServer(context);
	let running = true;
	let outFs;

	(instance as any).page = page;

	// Method to get output contents of the file system
	instance.getRawImage = () => {
		return outFs;
	}

	// Method to stop this instance
	instance.stop = async () => {
		// Grab the contents of fiel system
		outFs = await page.evaluate(function() {
			const entries: [string, string][] = [];

			for (let i = 0; i < localStorage.length; i++) {
				const key = localStorage.key(i);

				entries.push([key, localStorage.getItem(key)]);
			}

			return JSON.stringify(entries);
		});

		running = false;
		instance.emit("end");

		await Promise.all([
			// Closing the page
			page.close(),
			// Closing the server
			new Promise(resolve => server.close(resolve)),
		]);

		// Closing the browser
		await browser.close();
	}

	// Method to wait for status to come
	instance.waitForStatus = (status: string) => {
		return new Promise(resolve => {
			const listener = (newStatus: string) => {
				if (newStatus === status) {
					instance.off("status", listener);
					resolve();
				}
			}

			instance.on("status", listener);
		});
	}

	(async () => {
		while (running) {
			try {
				const ccStatus = await page.waitForSelector("#ccstatus", {state: "attached"});
				const status = await ccStatus.innerText();

				await page.evaluate("__clearStatus__()")

				instance.emit("status", status);
			} catch (error) {
				console.error(error);
				return;
			}
		}
	})().then()

	return instance;
}

async function startServer(context: Context): Promise<http.Server> {
	if (context.server) {
		return context.server;
	}

	const root = path.resolve(__dirname, "../../vendor/copy-cat");

	const server = createServer({
		root,
	});

	context.server = server;

	server.listen(context.options.args.port);

	const serverImpl: http.Server = typeof (server as any).server?.close === "function"
		? (server as any).server
		: server;

	server.close = function() {
		console.log("Server stopped gracefully");
		return serverImpl.close.apply(serverImpl, arguments);
	}

	console.log(`Copy-cat root is ${root}`);
	console.log(`Listening on http://localhost:${context.options.args.port}`);

	return server;
}

async function main(context: Context): Promise<ComputerCraftInstance> {
	let instance = await initializePage(context);

	await instance.page.goto(`http://localhost:${context.options.args.port}`);

	if (context.options.args.watch) {
		fs.watchFile(context.imageFileAbs, async () => {
			await instance.page.close();
			instance = await initializePage(context);

			await instance.page.goto(`http://localhost:${context.options.args.port}`);
		});
	}

	return instance;
}

export interface RunOptions {
	args?: {
		fs?: string;
		folder?: string;
		port: number;
		watch?: boolean;
	}
}

export function run(options: RunOptions): Promise<ComputerCraftInstance> {
	const {fs: imageFile, folder} = options.args;

	if (typeof imageFile === "string") {
		const imageFileAbs = resolvePath(imageFile);

		return main({
			imageFileAbs,
			options,
		});
	} else {
		const folderAbs = resolvePath(folder);

		const image = packImage({
			args: {
				folder: folderAbs,
				output: undefined,
				noEmit: true,
			}
		});

		return main({
			image,
			options,
		}).then(instance => {
			instance.on("end", () => {
				const imageRaw = instance.getRawImage();

				// Unpacking the image after the test is sun
				unpackImage({
					args: {
						image: imageRaw,
						output: folderAbs,
					}
				});
			});

			return instance;
		})
	}
}
