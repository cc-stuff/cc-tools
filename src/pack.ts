import {argv} from 'yargs';
import {assertIsString} from "./common/assertIsString";
import {packImage} from "./api";

const {folder, output} = argv;

assertIsString(folder);
assertIsString(output);

packImage({
	args: {
		folder,
		output,
	}
});
