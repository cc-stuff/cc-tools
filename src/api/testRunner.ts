import {bundleProject} from "./project";
import {resolvePath} from "../common/resolvePath";
import * as fs from "fs-extra";
import * as path from "path";
import {glob} from "glob";
import * as colors from "colors/safe"
import {run} from "./runner";

interface TestProject {
	testMatch?: string[];
}

enum TestDiffTokenStatus {
	In = "[+]",
	Out = "[-]",
	Same = "[=]",
}

type TestDiff = [string, TestDiffTokenStatus][];

interface TestOutputEntry {
	suiteName: string;
	testName: string;
	text: string;
	fail: boolean;
	diff?: TestDiff;
}

type TestOutput = TestOutputEntry[];

interface TestResultContext {
	totalTests: number;
	failedTests: number;
	output: TestOutput;
}

export interface TestOptions {
	outputLines?: string[];
	args?: {
		project: string;
		testsGrep?: string;
		suitsGrep?: string;
	}
}

function randomPort(): number {
	return 5000 + Math.round(Math.random() * 5000);
}

function displayTests(output: TestOutput, result: string[]): string[] {
	const bySuite: Record<string, Record<string, TestOutputEntry[]>> = {};

	for (const entry of output) {
		const byTest = bySuite[entry.suiteName] = bySuite[entry.suiteName] ?? {};
		const testOutputEntries = byTest[entry.testName] = byTest[entry.testName] ?? [];

		testOutputEntries.push(entry);
	}

	const bySuiteKeys = Object.keys(bySuite).sort();

	for (const suiteName of bySuiteKeys) {
		const byTestKeys = Object.keys(bySuite[suiteName]).sort();

		const suiteStartIndex = result.length;
		let failed = false;

		result.push(colors.bgGreen(colors.white("PASS")) + " " + colors.green(suiteName));

		for (const testName of byTestKeys) {
			const entries = bySuite[suiteName][testName];

			for (const entry of entries) {
				// TODO: Output levels
				if (entry.fail) {
					failed = true;

					result.push(colors.bold(colors.red(`    â€¢ ${testName}`)));
					result.push(colors.red(entry.text));
					result.push("");

					if (entry.diff) {
						for (const [line, status] of entry.diff) {
							switch (status) {
								case TestDiffTokenStatus.In:
									result.push(colors.green(status + " " + line));
									break;

								case TestDiffTokenStatus.Out:
									result.push(colors.red(status + " " + line));
									break;

								case TestDiffTokenStatus.Same:
									result.push(line);
									break;
							}
						}
					}
				}
			}

			if (failed) {
				result[suiteStartIndex] = (colors.bgRed(colors.white("FAIL")) + " " + colors.red(suiteName));
			}
		}
	}

	return result;
}

export async function test(options: TestOptions) {
	const projectAbs = resolvePath(options.args.project);
	const projectRoot = path.dirname(projectAbs);
	let outputLines: string[] = options.outputLines ?? [];

	// Loading project json
	const testProject: TestProject = JSON.parse(fs.readFileSync(projectAbs, "utf-8"));

	if (!Array.isArray(testProject.testMatch)) {
		throw new Error("No tests found! Please ensure that your project file contains a valid \"testMatch\" section.");
	}

	const startupFileName = path.join(projectRoot, "startup.lua");
	const commandFileName = path.join(projectRoot, ".command.json");
	const outputFileName = path.join(projectRoot, ".output.json");
	const statusFileName = path.join(projectRoot, ".ccstatus");

	if (fs.existsSync(startupFileName)) {
		throw new Error("No startup.lua in tests root allowed.")
	}

	const filesToDeleteAbs: string[] = [startupFileName, commandFileName, outputFileName, statusFileName];
	const filesWithTests: string[] = [];

	// Helper to remove temporary bundles
	function removeTempBundles() {
		for (const fileAbs of filesToDeleteAbs) {
			try {
				fs.unlinkSync(fileAbs);
			} catch (e) {
				console.error(e);
			}
		}
	}

	// Scanning for test files
	try {
		for (const match of testProject.testMatch) {
			const files = glob.sync(match, {cwd: projectRoot});

			// Globbing all the test files and building each and every near the original
			for (const entry of files) {
				const entryAbs = resolvePath(entry, projectRoot);
				const entryOut = entryAbs + ".tmp.bundle";

				filesToDeleteAbs.push(entryOut);
				filesWithTests.push(entryOut.replace(projectRoot, ""))

				bundleProject({
					args: {
						entry: resolvePath(entry, projectRoot),
						output: entryOut,
					}
				});
			}
		}

		// Preparing startup bundle
		bundleProject({
			args: {
				entry: resolvePath("../assets/test-runner.lua", __dirname),
				output: startupFileName,
			}
		});

		// Writing command.json
		fs.writeFileSync(commandFileName, JSON.stringify({
			suitsGrep: options.args.suitsGrep ?? ".*",
			testsGrep: options.args.testsGrep ?? ".*",
			files: filesWithTests,
		}));

		// Running tests
		const instance = await run({
			args: {
				folder: projectRoot,
				port: randomPort(),
			}
		});

		// Waiting for tests to end
		await instance.waitForStatus("ended");

		// Stopping the test instance
		await instance.stop();

		// Reading output.json
		const testResultRaw = fs.readFileSync(outputFileName, "utf-8");
		const testResult: TestResultContext = new Function("return " + testResultRaw)();

		if (testResult.totalTests === 0) {
			throw new Error("No tests found.");
		}

		// No failures, just display all run tests grouped by suite
		outputLines = displayTests(testResult.output, outputLines);

		outputLines.push("");
		outputLines.push(`Total tests: ${testResult.totalTests}`);
		outputLines.push(`Failed tests: ${testResult.failedTests}`);

		if (testResult.failedTests > 0) {
			throw new Error(`${testResult.failedTests} tests failed.`);
		}
	} catch (error) {
		removeTempBundles();

		throw error;
	}

	removeTempBundles();
}
