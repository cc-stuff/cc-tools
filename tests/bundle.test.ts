import * as fs from "fs";
import {bundleProject} from "../src/api";
import {clearDist} from "./utils";

describe("bundle", () => {

	beforeEach(() => {
		clearDist("./tests/assets/project1/dist")
		clearDist("./tests/assets/project2/dist")
	});

	test("project1 bundling", () => {
		bundleProject({
			projectFile: "./tests/assets/project1/project.json"
		});

		expect(
			fs.readFileSync("./tests/assets/project1/dist/project1.lua", "utf-8")
		).toMatchSnapshot();
	});

	test("project1 bundling with explicit args", () => {
		bundleProject({
			args: {
				entry: "./tests/assets/project1/index.lua",
				output: "./tests/assets/project1/dist/project1-explicit.lua",
			},
		});

		expect(
			fs.readFileSync("./tests/assets/project1/dist/project1-explicit.lua", "utf-8")
		).toMatchSnapshot();
	});

	test("project1 bundling with noEmit", () => {
		const bundleSource = bundleProject({
			args: {
				entry: "./tests/assets/project1/index.lua",
				noEmit: true,
			},
		});

		expect(bundleSource).toMatchSnapshot();
	});

	test("project2 bundling with references", () => {
		bundleProject({
			projectFile: "./tests/assets/project2/project.json"
		});

		expect(
			fs.readFileSync("./tests/assets/project2/dist/project2.lua", "utf-8")
		).toMatchSnapshot();

		expect(
			fs.readFileSync("./tests/assets/project2/dist/validator.lua", "utf-8")
		).toMatchSnapshot();
	});

});
