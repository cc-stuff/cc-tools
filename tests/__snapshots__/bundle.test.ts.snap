// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundle project1 bundling 1`] = `
"
local __files__ = {}

local function __require__(name)
	return __files__[name]
end

local function __define__(name, impl)
	__files__[name] = impl()
end


__define__('calc.lua', function ()

local function add(nA, nB)
    return nA + nB
end

return {
    add = add
}

end)

__define__('index.lua', function ()
local calc = __require__('calc.lua')
local calc2 = __require__('calc.lua')

print(calc.add(5, 2))

end)


__require__('index.lua')
"
`;

exports[`bundle project1 bundling with explicit args 1`] = `
"
local __files__ = {}

local function __require__(name)
	return __files__[name]
end

local function __define__(name, impl)
	__files__[name] = impl()
end


__define__('calc.lua', function ()

local function add(nA, nB)
    return nA + nB
end

return {
    add = add
}

end)

__define__('index.lua', function ()
local calc = __require__('calc.lua')
local calc2 = __require__('calc.lua')

print(calc.add(5, 2))

end)


__require__('index.lua')
"
`;

exports[`bundle project1 bundling with noEmit 1`] = `
"
local __files__ = {}

local function __require__(name)
	return __files__[name]
end

local function __define__(name, impl)
	__files__[name] = impl()
end


__define__('calc.lua', function ()

local function add(nA, nB)
    return nA + nB
end

return {
    add = add
}

end)

__define__('index.lua', function ()
local calc = __require__('calc.lua')
local calc2 = __require__('calc.lua')

print(calc.add(5, 2))

end)


__require__('index.lua')
"
`;

exports[`bundle project2 bundling with references 1`] = `
"
local __files__ = {}

local function __require__(name)
	return __files__[name]
end

local function __define__(name, impl)
	__files__[name] = impl()
end


__define__('index.lua', function ()
print(\\"It works!\\")

end)


__require__('index.lua')
"
`;

exports[`bundle project2 bundling with references 2`] = `
"
local __files__ = {}

local function __require__(name)
	return __files__[name]
end

local function __define__(name, impl)
	__files__[name] = impl()
end


__define__('is-number.lua', function ()

local function isNumber(a)
    return type(a) == \\"number\\"
end

return isNumber

end)

__define__('index.lua', function ()
local isNumber = __require__('is-number.lua')

return {
    assertIsNumber = function(a)
        assert(isNumber(a))
    end
}

end)


__require__('index.lua')
"
`;
